app/src/main/java/pl/imobitech/androidkurs/features/characters/presentation/model/CharacterDisplayable.kt

```diff
@@ -0,0 +1,33 @@
+package pl.imobitech.androidkurs.features.characters.presentation.model
+
+import pl.imobitech.androidkurs.features.characters.domain.model.Character
+import pl.imobitech.androidkurs.features.characters.domain.model.Origin
+import pl.imobitech.androidkurs.features.locations.domain.model.Location
+
+data class CharacterDisplayable(
+    val id: Int,
+    val name: String,
+    val status: String,
+    val species: String,
+    val type: String,
+    val gender: String,
+    val origin: Origin,
```

Comment:
Tutaj wystarczy jeszcze dodać `CharacterOriginDisplayable` i `LocationOriginDisplayable` zamiast tych modeli z warstwy domeny i będzie idealnie 👍

Reply:
Cześć, czy chodziło o modele z warstwy prezentacji: 
```
  val origin: CharacterOrigin,
  val location: CharacterLocation,
```
??

W konstruktorze mam też błąd pomylenia typów przy 

```
  origin = character.origin,
  location = character.location,
```
[https://ibb.co/hY3wRh1](url)


po usunięciu błędu importy wyglądałyby tak: 
```

import pl.imobitech.androidkurs.features.characters.domain.model.Character
import pl.imobitech.androidkurs.features.characters.domain.model.CharacterOrigin
```

Reply:
W zasadzie bardzo blisko. Chodzi o to, że każda z warstw takiej architektury powinna mieć swoje modele - przy większych projektach daje to sporą elastyczność. Teraz dla warstwy danych obiekt `CharacterRemote` zawiera w sobie obiekty `CharacterOriginRemote` oraz `CharacterLocationRemote`. Odpowiednik tego obiektu dla warstwy domeny - `Character` zawiera kolejno `CharacterOrigin` i `CharacterLocation`. Teraz chodzi o to, żeby dla obiektu z warstwy prezentacji, czyli `CharacterDisplayable` zrobić to samo. W tym przypadku taki obiekt powinien zawierać obiekty klas `CharacterOriginDisplayable` i `CharacterLocationDisplayable`, odpowiadają analogicznym obiektom z innych warstw. Mogłoby to wyglądać tak: 
```
data class CharacterDisplayable(
    val id: Int,
    val name: String,
    val status: String,
    val species: String,
    val type: String,
    val gender: String,
    val origin: CharacterOriginDisplayable,
    val location: CharacterLocationDisplayable,
    val episode: List<String>,
    val image: String,
    val url: String
) {
    constructor(character: Character) : this(
        id = character.id,
        name = character.name,
        status = character.status,
        species = character.species,
        type = character.type,
        gender = character.gender,
        origin = CharacterOriginDisplayable(character.origin),
        location = CharacterLocationDisplayable(character.location),
        image = character.image,
        episode = character.episode,
        url = character.url
    )
}

data class CharacterOriginDisplayable(
    val name: String,
    val url: String
) {
    constructor(origin: CharacterOrigin) : this(
        name = origin.name,
        url = origin.url
    )
}

data class CharacterLocationDisplayable(
    val name: String,
    val url: String
) {
    constructor(location: CharacterLocation) : this(
        name = location.name,
        url = location.url
    )
} 
```